File: backend/app/schemas/team.py
# ABOUTME: Pydantic schemas for team data validation and serialization
# ABOUTME: Defines request/response models for team API endpoints

from datetime import datetime

from pydantic import BaseModel, Field

from .player import PlayerResponse


class TeamBase(BaseModel):
    """Base team schema with common fields"""

    player1_id: int = Field(..., description="First team member ID")
    player2_id: int = Field(..., description="Second team member ID")


class TeamCreate(TeamBase):
    """Schema for creating a new team"""

    pass


class TeamResponse(TeamBase):
    """Schema for team response data"""

    id: int
    trueskill_mu: float
    trueskill_sigma: float
    games_played: int
    wins: int
    losses: int
    win_percentage: float
    conservative_rating: float
    player_names: str
    is_active: bool
    created_at: datetime
    updated_at: datetime | None
    player1: PlayerResponse
    player2: PlayerResponse

    class Config:
        from_attributes = True


class TeamListResponse(BaseModel):
    """Schema for paginated team list response"""

    teams: list[TeamResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class TeamGameBase(BaseModel):
    """Base team game schema with common fields"""

    team1_id: int = Field(..., description="First team ID")
    team2_id: int = Field(..., description="Second team ID")
    winner_team_id: int = Field(..., description="Winner team ID")


class TeamGameCreate(TeamGameBase):
    """Schema for creating a new team game"""

    pass


class TeamGameResponse(TeamGameBase):
    """Schema for team game response data"""

    id: int
    created_at: datetime
    team1: TeamResponse
    team2: TeamResponse
    winner_team: TeamResponse

    class Config:
        from_attributes = True


class TeamGameListResponse(BaseModel):
    """Schema for paginated team game list response"""

    team_games: list[TeamGameResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class TeamFormationRequest(BaseModel):
    """Schema for requesting team formation from 4 players"""

    team1_player1_id: int = Field(..., description="Team 1 - First player ID")
    team1_player2_id: int = Field(..., description="Team 1 - Second player ID")
    team2_player1_id: int = Field(..., description="Team 2 - First player ID")
    team2_player2_id: int = Field(..., description="Team 2 - Second player ID")
    winner_team: int = Field(..., ge=1, le=2, description="Winning team (1 or 2)")

File: frontend/src/types/team.ts
// ABOUTME: TypeScript types for Team and TeamGame data matching backend Pydantic schemas
// ABOUTME: Ensures type safety between frontend and backend team functionality

import type { Player } from './player'

export interface Team {
  id: number
  player1_id: number
  player2_id: number
  trueskill_mu: number
  trueskill_sigma: number
  games_played: number
  wins: number
  losses: number
  created_at: string
  updated_at?: string
  is_active: boolean
  win_percentage: number
  conservative_rating: number
  player_names: string
  player1: Player
  player2: Player
}

export interface TeamCreate {
  player1_id: number
  player2_id: number
}

export interface TeamGame {
  id: number
  team1_id: number
  team2_id: number
  winner_team_id: number
  created_at: string
  team1: Team
  team2: Team
  winner_team: Team
  loser_team_id: number
  loser_team: Team
}

export interface TeamGameCreate {
  team1_id: number
  team2_id: number
  winner_team_id: number
}

export interface TeamFormationRequest {
  player1_id: number
  player2_id: number
  player3_id: number
  player4_id: number
  winner_team: 1 | 2
}

export interface TeamListResponse {
  teams: Team[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface TeamGameListResponse {
  team_games: TeamGame[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface TeamSearchResponse {
  team: Team | null
}

export interface MatchQuality {
  quality: number
  description: string
}
